x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -202,17 +219,42 @@ func (sm *StorageManager) Size() (lsm, vlog int64) {
 	// Also remember to update
 	// - x-pack/apm-server/sampling/processor.go:CollectMonitoring
 	// - systemtest/benchtest/expvar/metrics.go
-	return int64(sm.DiskUsage()), 0
+	return int64(sm.dbSize()), 0
 }
 
-// DiskUsage returns the disk usage of databases in bytes.
-func (sm *StorageManager) DiskUsage() uint64 {
+// dbSize returns the disk usage of databases in bytes.
+func (sm *StorageManager) dbSize() uint64 {
 	// pebble DiskSpaceUsage overhead is not high, but it adds up when performed per-event.
-	return sm.cachedDiskUsage.Load()
+	return sm.cachedDBSize.Load()
+}
+
+// dbStorageLimit returns the configured limit of the database size in bytes.
+func (sm *StorageManager) dbStorageLimit() uint64 {
+	return sm.storageLimit.Load()
 }
 
 func (sm *StorageManager) updateDiskUsage() {
-	sm.cachedDiskUsage.Store(sm.eventDB.Metrics().DiskSpaceUsage() + sm.decisionDB.Metrics().DiskSpaceUsage())
+	sm.cachedDBSize.Store(sm.eventDB.Metrics().DiskSpaceUsage() + sm.decisionDB.Metrics().DiskSpaceUsage())
+
+	usage, err := vfs.Default.GetDiskUsage(sm.storageDir)
+	if err != nil {
+		sm.rateLimitedLogger.With(logp.Error(err)).Warn("failed to get disk usage")
```

Comment:
what is the fallback in this case?

Reply:
In the current implementation it effectively becomes unlimited. It is by design because of how they are 2 separate checkers and RW. But I _think_ it is possible to overwrite `dbStorageLimit` to a fallback value if we get an error, so that storage_limit checker takes over.

Reply:
Added fallback handling. A bit complex, but should cover all cases. All of it is written with the assumption that if GetDiskUsage ever returns an error, it will always return an error.